src/OpenLoco/src/GameCommands/RemoveTown.cpp

```diff
@@ -0,0 +1,131 @@
+#include "Audio/Audio.h"
+#include "Economy/Expenditures.h"
+#include "GameCommands.h"
+#include "Localisation/StringIds.h"
+#include "Map/BuildingElement.h"
+#include "Map/RoadElement.h"
+#include "Map/TileElement.h"
+#include "Map/TileLoop.hpp"
+#include "Map/TileManager.h"
+#include "Objects/ObjectManager.h"
+#include "S5/S5.h"
+#include "ViewportManager.h"
+#include "World/StationManager.h"
+#include "World/TownManager.h"
+
+using namespace OpenLoco::Interop;
+using namespace OpenLoco::World;
+
+namespace OpenLoco::GameCommands
+{
+    static uint32_t removeTown(const TownRemovalArgs& args, const uint8_t flags)
+    {
+        for (auto& station : StationManager::stations())
+        {
+            if (station.town == args.townId)
+            {
+                setErrorText(StringIds::all_stations_near_this_town_must_be_removed_first);
+                return FAILURE;
+            }
+        }
+
+        if (!(flags & Flags::apply))
+        {
+            return 0;
+        }
+
+        // NB: vanilla did not set an expenditute type
+        GameCommands::setExpenditureType(ExpenditureType::Construction);
+
+        // Iterate over the entire map to find town tiles
+        // TODO: can't we do better than vanilla for this? e.g. a radius around the town centre?
+        TilePosRangeView tileLoop{ TilePos2{ 0, 0 }, TilePos2{ kMapColumns, kMapRows } };
+
+        for (auto& tilePos : tileLoop)
+        {
+            auto tile = TileManager::get(tilePos);
+
+            for (auto& element : tile)
```

Comment:
This loop is crashing for reasons unclear to me.
```
* thread #1, name = 'OpenLoco', stop reason = signal SIGSEGV: invalid address (fault address: 0x1)
    frame #0: 0x012274b3 OpenLoco`OpenLoco::World::Tile::end(this=0xffffcf4c) at Tile.cpp:58:18
   55           do
   56           {
   57               el++;
-> 58           } while (!(el - 1)->isLast());
   59           return el;
   60       }
   61  
(lldb) bt
* thread #1, name = 'OpenLoco', stop reason = signal SIGSEGV: invalid address (fault address: 0x1)
  * frame #0: 0x012274b3 OpenLoco`OpenLoco::World::Tile::end(this=0xffffcf4c) at Tile.cpp:58:18
    frame #1: 0x011f6856 OpenLoco`OpenLoco::GameCommands::removeTown(OpenLoco::Interop::registers&) at RemoveTown.cpp:48:34
    frame #2: 0x011f67ca OpenLoco`OpenLoco::GameCommands::removeTown(regs=0xffffcff0) at RemoveTown.cpp:129:30
    frame #3: 0x011f4278 OpenLoco`OpenLoco::GameCommands::loc_4313C6(int, const OpenLoco::Interop::registers &) at GameCommands.cpp:254:39
    frame #4: 0x011f4266 OpenLoco`OpenLoco::GameCommands::loc_4313C6(esi=50, regs=0xffffd090) at GameCommands.cpp:316:32
    frame #5: 0x011f453c OpenLoco`OpenLoco::GameCommands::doCommand(command=removeTown, regs=0xffffd090) at GameCommands.cpp:206:32
    frame #6: 0x012dd315 OpenLoco`OpenLoco::Ui::Windows::Town::Town::onMouseUp(OpenLoco::Ui::Window &, OpenLoco::Ui::WidgetIndex_t) [inlined] unsigned int OpenLoco::GameCommands::doCommand<OpenLoco::GameCommands::TownRemovalArgs>(args=<unavailable>, flags='\x01') at GameCommands.h:138:25
    frame #7: 0x012dd2dd OpenLoco`OpenLoco::Ui::Windows::Town::Town::onMouseUp(self=0x01139b72, widgetIndex='\v') at TownWindow.cpp:220:48
    frame #8: 0x01201bb0 OpenLoco`OpenLoco::Input::handleMouse(short, short, OpenLoco::Input::MouseButton) at MouseInput.cpp:1527:38
    frame #9: 0x01201b51 OpenLoco`OpenLoco::Input::handleMouse(x=946, y=110, button=leftReleased) at MouseInput.cpp:447:35
    frame #10: 0x01263759 OpenLoco`OpenLoco::Ui::handleInput() at Ui.cpp:868:35
    frame #11: 0x0124e4e6 OpenLoco`OpenLoco::tick() at OpenLoco.cpp:406:24
    frame #12: 0x011a3686 OpenLoco`OpenLoco::main(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>>&&) at OpenLoco.cpp:1038:17
    frame #13: 0x01176ac7 OpenLoco`main at Main.cpp:28:26
    frame #14: 0xf741f6e9 libc.so.6`___lldb_unnamed_symbol3437 + 121
    frame #15: 0xf741f7ad libc.so.6`__libc_start_main + 141
    frame #16: 0x011abefb OpenLoco`_start + 43
```

Reply:
Perhaps this is due me skipping the tile pointer assignments to `0x00F00158`? I thought these could be omitted due to our tile loop not using them.